src/src/com/microsoft/aad/adal/HttpWebRequest.java

```diff
@@ -107,107 +89,103 @@ private void setupConnection() {
-                    // If it does not get the error stream, it will return
-                    // exception in the httpresponse
-                    responseStream = mConnection.getErrorStream();
-                    mException = ex;
+                // Continue to read from stream if not cancelled and not EOF
+                while ((bytesRead = responseStream.read(buffer)) > 0) {
+                    byteStream.write(buffer, 0, bytesRead);
                 }
 
-                // GET request should read status after getInputStream to make
-                // this work for different SDKs
-                getStatusCode(response);
-
-                if (responseStream != null) {
-
-                    ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
-                    byte[] buffer = new byte[4096];
-                    int bytesRead = -1;
-
-                    // Continue to read from stream if not cancelled and not EOF
-                    while ((bytesRead = responseStream.read(buffer)) > 0) {
-                        byteStream.write(buffer, 0, bytesRead);
-                    }
-
-                    responseBody = byteStream.toByteArray();
-                }
+                responseBody = byteStream.toByteArray();
+            }
 
-                // It will only run in debugger and set from outside for testing
-                if (android.os.Debug.isDebuggerConnected() && sDebugSimulateDelay > 0) {
-                    // sleep background thread in debugging mode
-                    Logger.v(TAG, "Sleeping to simulate slow network response");
+            // It will only run in debugger and set from outside for testing
+            if (Debug.isDebuggerConnected() && sDebugSimulateDelay > 0) {
+                // sleep background thread in debugging mode
+                Logger.v(TAG, "Sleeping to simulate slow network response");
+                try {
                     Thread.sleep(sDebugSimulateDelay);
+                } catch (InterruptedException e) {
+                    Logger.v(TAG, "Thread.sleep got interrupted exception " + e);
                 }
-
-                Logger.v(TAG, "Response is received");
-                response.setBody(responseBody);
-                response.setResponseHeaders(mConnection.getHeaderFields());
-            } catch (InterruptedException e) {
-                Logger.v(TAG, "Thread.sleep got interrupted exception " + e);
-            } catch (IOException e ) {
-                Logger.e(TAG, "IOException:" + e.getMessage(), " Method:" + mRequestMethod,
-                        ADALError.SERVER_ERROR, e);
-                mException = e;
-            } finally {
-                mConnection.disconnect();
-                mConnection = null;
             }
+
+            Logger.v(TAG, "Response is received");
+            response = new HttpWebResponse(statusCode, responseBody, connection.getHeaderFields());
+        } finally {
+            connection.disconnect();
```

Comment:
Hm. I think I know why we setting connection reuse headers :-) Documentation says to avoid calling this method as it may kill other connections. I don't think we should change this right now, but we should consider removing `System.setProperty("http.keepAlive", "false");` as well as `connection.disconnect();`